Python/compile.c

```diff
@@ -2274,9 +2496,48 @@ compiler_class(struct compiler *c, stmt_ty s)
     ADDOP_LOAD_CONST(c, loc, s->v.ClassDef.name);
 
     /* 5. generate the rest of the code for the call */
-    RETURN_IF_ERROR(compiler_call_helper(c, loc, 2,
-                                         s->v.ClassDef.bases,
-                                         s->v.ClassDef.keywords));
+
+    if (is_generic) {
+        _Py_DECLARE_STR(type_params, ".type_params");
+        _Py_DECLARE_STR(generic_base, ".generic_base");
+        RETURN_IF_ERROR(compiler_nameop(c, loc, &_Py_STR(type_params), Load));
+        ADDOP_I(c, loc, CALL_INTRINSIC_1, INTRINSIC_SUBSCRIPT_GENERIC);
+        RETURN_IF_ERROR(compiler_nameop(c, loc, &_Py_STR(generic_base), Store));
+
+        Py_ssize_t original_len = asdl_seq_LEN(s->v.ClassDef.bases);
+        asdl_expr_seq *bases = _Py_asdl_expr_seq_new(
+            original_len + 1, c->c_arena);
+        for (Py_ssize_t i = 0; i < original_len; i++) {
+            asdl_seq_SET(bases, i, asdl_seq_GET(s->v.ClassDef.bases, i));
+        }
+        expr_ty name_node = _PyAST_Name(
+            &_Py_STR(generic_base), Load,
+            loc.lineno, loc.col_offset, loc.end_lineno, loc.end_col_offset, c->c_arena
+        );
+        if (name_node == NULL) {
+            return ERROR;
+        }
+        asdl_seq_SET(bases, original_len, name_node);
+        RETURN_IF_ERROR(compiler_call_helper(c, loc, 2,
+                                             bases,
+                                             s->v.ClassDef.keywords));
```

Comment:
Will try this out, but just a drive-by comment that it's a LOAD_FAST, not a load to a dict:

```
Disassembly of <code object <generic parameters of X> at 0x105621690, file "<dis>", line 1>:
              0 MAKE_CELL                2 (.type_params)

  1           2 RESUME                   0
              4 LOAD_CONST               0 ('T')
              6 CALL_INTRINSIC_1         7 (INTRINSIC_TYPEVAR)
              8 COPY                     1
             10 STORE_FAST               0 (T)
             12 BUILD_TUPLE              1
             14 STORE_DEREF              2 (.type_params)
             16 PUSH_NULL
             18 LOAD_BUILD_CLASS
             20 LOAD_CLOSURE             2 (.type_params)
             22 BUILD_TUPLE              1
             24 LOAD_CONST               1 (<code object X at 0x105645e40, file "<dis>", line 1>)
             26 MAKE_FUNCTION            8 (closure)
             28 LOAD_CONST               2 ('X')
             30 LOAD_DEREF               2 (.type_params)
             32 CALL_INTRINSIC_1        10 (INTRINSIC_SUBSCRIPT_GENERIC)
             34 STORE_FAST               1 (.generic_base)
             36 LOAD_FAST                1 (.generic_base)
             38 CALL                     3
             46 RETURN_VALUE
```

Reply:
Thinking about it more, your suggestion is probably not simpler. We can't visit the bases in order `1, ..., n, 0` as you suggest, because we need to visit them in order, in case evaluation of any of the bases throws an error. I initially had a version of this code that passed an extra argument to `compiler_call_helper` to tell it where to insert the generic bases, but that didn't work for the "slow path" in that function (`ex_call`), where we generate a `CALL_FUNCTION_EX` opcode. After discussion with Irit, I switched to the approach in the current code. This approach avoids complicating the code for generating a call (which is already quite complicated!) at the cost of some manual work only for generics.

We need to be able to compile cases like `class X[T](A, B, c=3, d=4, **e): pass`, in addition to the simple cases where the stack manipulation is easy.

Reply:
I think handling keyword arguments and all that is no trouble at all, that can still be fully deferred to `compiler_call_helper` without needing to modify it; it's just a question of how many on-stack positional args we send to `compiler_call_helper` vs positional args we send as an AST sequence.

But the need to evaluate the bases in order is a good call, and does make things slightly more complicated. You'd have to instead evaluate the bases in order and use a `SWAP 2` after evaluating each base, to keep bubbling the generic base up to the top of the stack so it is the last base.

I still think this ends up simpler (you have to create an empty AST sequence to send as `args` to `compiler_call_helper`, but that's simpler than creating a Name node and adding it to a new length+1 sequence as you are doing now). And I still think I would prefer it, because in the common case of no extra bases, it results in the simplest possible bytecode with no extraneous loads or stores. But it's up to you.

Reply:
That makes sense, I didn't realize that `compiler_call_helper` supports arguments that are already on the stack. We could indeed push all the args onto the stack manually. I think the check against `STACK_USE_GUIDELINE` in compiler_call_helper would have to be changed to incorporate `n`. I'll give this a try after finishing your other comments.

Reply:
I realized it's even more complicated. What if we have this:

```python
class C[T](*bases): pass
```

Then we can't just use `compiler_call_helper`'s already-pushed-args support; we'd need to duplicate the code inside `compiler_call_helper` that unpacks the `*args`.

The current implementation supports this:

```
>>> class A: pass
... 
>>> bases = (A,)
>>> class B[T](*bases): pass
... 
>>> B.__orig_bases__
(<class '__main__.A'>, typing.Generic[T])
```

Reply:
Ah indeed. That makes this idea too complicated. Thanks for looking into it!